File: crates/fortress-go/audit.go
package fortress

import (
	"encoding/json"
	"sync"
	"time"
)

// AuditLogger manages audit logging
type AuditLogger struct {
	config  *Config
	entries []*AuditEntry
	mu      sync.RWMutex
}

// NewAuditLogger creates a new audit logger
func NewAuditLogger(config *Config) *AuditLogger {
	return &AuditLogger{
		config:  config,
		entries: make([]*AuditEntry, 0),
	}
}

// Log logs an audit entry
func (a *AuditLogger) Log(entry *AuditEntry) error {
	if !a.config.Audit.Enabled {
		return nil
	}

	a.mu.Lock()
	defer a.mu.Unlock()

	// Generate ID if not provided
	if entry.ID == "" {
		id, err := GenerateRandomString(16)
		if err != nil {
			return err
		}
		entry.ID = id
	}

	// Set timestamp if not provided
	if entry.Timestamp.IsZero() {
		entry.Timestamp = time.Now()
	}

	a.entries = append(a.entries, entry)

	// Batch flush if needed
	if len(a.entries) >= a.config.Audit.BatchSize {
		return a.Flush()
	}

	return nil
}

// LogSuccess logs a successful operation
func (a *AuditLogger) LogSuccess(tenantID, userID, action string, resource *Resource, metadata map[string]interface{}) error {
	return a.Log(&AuditEntry{
		TenantID: tenantID,
		UserID:   userID,
		Action:   action,
		Resource: *resource,
		Result:   "success",
		Metadata: metadata,
	})
}

// LogFailure logs a failed operation
func (a *AuditLogger) LogFailure(tenantID, userID, action string, resource *Resource, errorMsg string, metadata map[string]interface{}) error {
	return a.Log(&AuditEntry{
		TenantID: tenantID,
		UserID:   userID,
		Action:   action,
		Resource: *resource,
		Result:   "failure",
		Error:    errorMsg,
		Metadata: metadata,
	})
}

// Query queries audit entries with optional filters
func (a *AuditLogger) Query(options *AuditQueryOptions) ([]*AuditEntry, error) {
	a.mu.RLock()
	defer a.mu.RUnlock()

	entries := make([]*AuditEntry, len(a.entries))
	copy(entries, a.entries)

	// Filter by tenant
	if options != nil && options.TenantID != "" {
		var filtered []*AuditEntry
		for _, entry := range entries {
			if entry.TenantID == options.TenantID {
				filtered = append(filtered, entry)
			}
		}
		entries = filtered
	}

	// Filter by user
	if options != nil && options.UserID != "" {
		var filtered []*AuditEntry
		for _, entry := range entries {
			if entry.UserID == options.UserID {
				filtered = append(filtered, entry)
			}
		}
		entries = filtered
	}

	// Filter by action
	if options != nil && options.Action != "" {
		var filtered []*AuditEntry
		for _, entry := range entries {
			if entry.Action == options.Action {
				filtered = append(filtered, entry)
			}
		}
		entries = filtered
	}

	// Filter by resource type
	if options != nil && options.ResourceType != "" {
		var filtered []*AuditEntry
		for _, entry := range entries {
			if entry.Resource.Type == options.ResourceType {
				filtered = append(filtered, entry)
			}
		}
		entries = filtered
	}

	// Filter by result
	if options != nil && options.Result != "" {
		var filtered []*AuditEntry
		for _, entry := range entries {
			if entry.Result == options.Result {
				filtered = append(filtered, entry)
			}
		}
		entries = filtered
	}

	// Filter by date range
	if options != nil && options.StartDate != nil {
		var filtered []*AuditEntry
		for _, entry := range entries {
			if entry.Timestamp.After(*options.StartDate) || entry.Timestamp.Equal(*options.StartDate) {
				filtered = append(filtered, entry)
			}
		}
		entries = filtered
	}

	if options != nil && options.EndDate != nil {
		var filtered []*AuditEntry
		for _, entry := range entries {
			if entry.Timestamp.Before(*options.EndDate) || entry.Timestamp.Equal(*options.EndDate) {
				filtered = append(filtered, entry)
			}
		}
		entries = filtered
	}

	// Sort by timestamp (newest first)
	for i := 0; i < len(entries)-1; i++ {
		for j := i + 1; j < len(entries); j++ {
			if entries[i].Timestamp.Before(entries[j].Timestamp) {
				entries[i], entries[j] = entries[j], entries[i]
			}
		}
	}

	// Apply pagination
	if options != nil && options.Offset > 0 {
		if options.Offset >= len(entries) {
			return []*AuditEntry{}, nil
		}
		entries = entries[options.Offset:]
	}

	if options != nil && options.Limit > 0 && options.Limit < len(entries) {
		entries = entries[:options.Limit]
	}

	return entries, nil
}

// GetStats returns audit statistics
func (a *AuditLogger) GetStats(options *AuditStatsOptions) (*AuditStats, error) {
	entries, err := a.Query(&AuditQueryOptions{
		TenantID: options.TenantID,
		UserID:   options.UserID,
		Action:   options.Action,
	})
	if err != nil {
		return nil, NewAuditError("failed to query entries for stats", err)
	}

	stats := &AuditStats{
		TotalEntries:         len(entries),
		SuccessfulOperations: 0,
		FailedOperations:     0,
		UniqueUsers:          0,
		UniqueResources:      0,
		UniqueActions:        make(map[string]bool),
		Actions:              make(map[string]int),
		HourlyStats:          make([]int, 24),
	}

	// Track unique users and resources
	uniqueUsers := make(map[string]bool)
	uniqueResources := make(map[string]bool)

	if len(entries) > 0 {
		stats.TimeRange.Start = entries[len(entries)-1].Timestamp
		stats.TimeRange.End = entries[0].Timestamp
	}

	for _, entry := range entries {
		if entry.Result == "success" {
			stats.SuccessfulOperations++
		} else {
			stats.FailedOperations++
		}

		if entry.UserID != "" {
			if !uniqueUsers[entry.UserID] {
				uniqueUsers[entry.UserID] = true
				stats.UniqueUsers++
			}
		}

		if entry.Resource.ID != "" {
			if !uniqueResources[entry.Resource.ID] {
				uniqueResources[entry.Resource.ID] = true
				stats.UniqueResources++
			}
		}

		stats.UniqueActions[entry.Action] = true
		stats.Actions[entry.Action]++
	}

	return stats, nil
}

// Flush flushes audit entries to storage
func (a *AuditLogger) Flush() error {
	a.mu.Lock()
	defer a.mu.Unlock()

	if len(a.entries) == 0 {
		return nil
	}

	// TODO: Implement actual storage flush
	// For now, just clear the entries
	a.entries = make([]*AuditEntry, 0)
	return nil
}

// Cleanup removes old audit entries based on retention policy
func (a *AuditLogger) Cleanup() error {
	a.mu.Lock()
	defer a.mu.Unlock()

	retentionDays := a.config.Audit.RetentionDays
	if retentionDays <= 0 {
		return nil
	}

	cutoff := time.Now().AddDate(0, 0, -retentionDays)
	var remaining []*AuditEntry

	for _, entry := range a.entries {
		if entry.Timestamp.After(cutoff) {
			remaining = append(remaining, entry)
		}
	}

	a.entries = remaining
	return nil
}

// GetEntryCount returns the number of entries
func (a *AuditLogger) GetEntryCount() int {
	a.mu.RLock()
	defer a.mu.RUnlock()
	return len(a.entries)
}

// Clear clears all audit entries
func (a *AuditLogger) Clear() error {
	a.mu.Lock()
	defer a.mu.Unlock()

	a.entries = make([]*AuditEntry, 0)
	return nil
}

// ExportToJSON exports audit entries to JSON
func (a *AuditLogger) ExportToJSON() ([]byte, error) {
	a.mu.RLock()
	defer a.mu.RUnlock()

	return json.MarshalIndent(a.entries, "", "  ")
}

// ImportFromJSON imports audit entries from JSON
func (a *AuditLogger) ImportFromJSON(data []byte) error {
	var entries []*AuditEntry
	err := json.Unmarshal(data, &entries)
	if err != nil {
		return NewAuditError("failed to import from JSON", err)
	}

	a.mu.Lock()
	defer a.mu.Unlock()

	a.entries = append(a.entries, entries...)
	return nil
}

File: crates/fortress-js/src/audit.ts
/**
 * Audit logging functionality for Fortress JavaScript/TypeScript SDK
 */

import { FortressConfig } from './config';
import { FortressError } from './error';
import { AuditEntry, Resource } from './types';

/**
 * Audit logger class
 */
export class AuditLogger {
  private config: FortressConfig;
  private entries: AuditEntry[];

  constructor(config: FortressConfig) {
    this.config = config;
    this.entries = [];
  }

  /**
   * Log an audit entry
   */
  async log(entry: Omit<AuditEntry, 'id' | 'timestamp'>): Promise<void> {
    try {
      if (!this.config.auditConfig.enabled) {
        return;
      }

      const auditEntry: AuditEntry = {
        id: this.generateId(),
        timestamp: new Date(),
        ...entry,
      };

      this.entries.push(auditEntry);

      // Batch flush if needed
      if (this.entries.length >= this.config.auditConfig.batch_size!) {
        await this.flush();
      }
    } catch (error) {
      throw FortressError.audit('Failed to log audit entry', error as Error);
    }
  }

  /**
   * Log successful operation
   */
  async logSuccess(
    tenantId: string,
    userId: string,
    action: string,
    resource: Resource,
    metadata?: any
  ): Promise<void> {
    await this.log({
      tenant_id: tenantId,
      user_id: userId,
      action,
      resource,
      result: 'success',
      metadata,
    });
  }

  /**
   * Log failed operation
   */
  async logFailure(
    tenantId: string,
    userId: string,
    action: string,
    resource: Resource,
    error: string,
    metadata?: any
  ): Promise<void> {
    await this.log({
      tenant_id: tenantId,
      user_id: userId,
      action,
      resource,
      result: 'failure',
      error,
      metadata,
    });
  }

  /**
   * Query audit entries
   */
  async query(options: AuditQueryOptions = {}): Promise<AuditEntry[]> {
    try {
      let entries = [...this.entries];

      // Filter by tenant
      if (options.tenantId) {
        entries = entries.filter(entry => entry.tenant_id === options.tenantId);
      }

      // Filter by user
      if (options.userId) {
        entries = entries.filter(entry => entry.user_id === options.userId);
      }

      // Filter by action
      if (options.action) {
        entries = entries.filter(entry => entry.action === options.action);
      }

      // Filter by resource type
      if (options.resourceType) {
        entries = entries.filter(entry => entry.resource.type === options.resourceType);
      }

      // Filter by result
      if (options.result) {
        entries = entries.filter(entry => entry.result === options.result);
      }

      // Filter by date range
      if (options.startDate) {
        entries = entries.filter(entry => entry.timestamp >= options.startDate!);
      }

      if (options.endDate) {
        entries = entries.filter(entry => entry.timestamp <= options.endDate!);
      }

      // Sort by timestamp (newest first)
      entries.sort((a, b) => b.timestamp.getTime() - a.timestamp.getTime());

      // Apply pagination
      if (options.offset) {
        entries = entries.slice(options.offset);
      }

      if (options.limit) {
        entries = entries.slice(0, options.limit);
      }

      return entries;
    } catch (error) {
      throw FortressError.audit('Failed to query audit entries', error as Error);
    }
  }

  /**
   * Get audit statistics
   */
  async getStats(options?: AuditStatsOptions): Promise<AuditStats> {
    try {
      const entries = await this.query(options);
      
      const stats: AuditStats = {
        totalEntries: entries.length,
        successfulOperations: entries.filter(e => e.result === 'success').length,
        failedOperations: entries.filter(e => e.result === 'failure').length,
        uniqueUsers: new Set(entries.map(e => e.user_id).filter(Boolean)).size,
        uniqueActions: new Set(entries.map(e => e.action)).size,
        timeRange: {
          start: entries.length > 0 ? entries[entries.length - 1].timestamp : new Date(),
          end: entries.length > 0 ? entries[0].timestamp : new Date(),
        },
      };

      return stats;
    } catch (error) {
      throw FortressError.audit('Failed to get audit statistics', error as Error);
    }
  }

  /**
   * Flush audit entries to storage
   */
  async flush(): Promise<void> {
    try {
      if (this.entries.length === 0) {
        return;
      }

      // Create a snapshot of entries to flush
      const entriesToFlush = [...this.entries];
      
      // Store audit entries in batches to avoid memory issues
      const batchSize = 100;
      for (let i = 0; i < entriesToFlush.length; i += batchSize) {
        const batch = entriesToFlush.slice(i, i + batchSize);
        const batchKey = `audit_batch_${Date.now()}_${i}`;
        
        // Serialize the batch
        const serializedBatch = JSON.stringify(batch);
        const data = new TextEncoder().encode(serializedBatch);
        
        // Store using the configured storage backend
        // Note: In a real implementation, this would use the Fortress storage backend
        // For now, we'll store in localStorage if available, otherwise memory
        if (typeof localStorage !== 'undefined') {
          try {
            localStorage.setItem(batchKey, serializedBatch);
          } catch (error) {
            console.warn('Failed to store audit batch in localStorage:', error);
            // Fallback to memory storage
            this.storeInMemory(batchKey, data);
          }
        } else {
          // Node.js or environment without localStorage
          this.storeInMemory(batchKey, data);
        }
      }
      
      // Clear flushed entries from memory
      this.entries = this.entries.slice(entriesToFlush.length);
      
    } catch (error) {
      throw FortressError.audit('Failed to flush audit entries', error as Error);
    }
  }

  /**
   * Store data in memory as fallback
   */
  private storeInMemory(key: string, data: Uint8Array): void {
    // Simple in-memory storage fallback
    // In a real implementation, this would use the Fortress storage backend
    if (!(globalThis as any)._fortressAuditStorage) {
      (globalThis as any)._fortressAuditStorage = new Map();
    }
    (globalThis as any)._fortressAuditStorage.set(key, data);
  }

  /**
   * Clear old audit entries based on retention policy
   */
  async cleanup(): Promise<void> {
    try {
      const retentionDays = this.config.auditConfig.retention_days || 30;
      const cutoffDate = new Date();
      cutoffDate.setDate(cutoffDate.getDate() - retentionDays);

      this.entries = this.entries.filter(entry => entry.timestamp >= cutoffDate);
    } catch (error) {
      throw FortressError.audit('Failed to cleanup audit entries', error as Error);
    }
  }

  /**
   * Generate unique ID for audit entries
   */
  private generateId(): string {
    return Math.random().toString(36).substr(2, 9) + Date.now().toString(36);
  }
}

export interface AuditQueryOptions {
  tenantId?: string;
  userId?: string;
  action?: string;
  resourceType?: string;
  result?: 'success' | 'failure';
  startDate?: Date;
  endDate?: Date;
  offset?: number;
  limit?: number;
}

export interface AuditStatsOptions extends AuditQueryOptions {}

export interface AuditStats {
  totalEntries: number;
  successfulOperations: number;
  failedOperations: number;
  uniqueUsers: number;
  uniqueActions: number;
  timeRange: {
    start: Date;
    end: Date;
  };
}
